ferry.go

```diff
@@ -627,6 +663,86 @@ func (f *Ferry) SerializeStateToJSON() (string, error) {
 	return string(stateBytes), err
 }
 
+func (f *Ferry) Progress() *Progress {
```

Comment:
Is it worth breaking up any of the below functionality into separate methods? Seems like there is a lot going on here.

Reply:
I personally rather get the ugliness over with all in one method rather than putting it in multiple places.